### PR TITLE
rewrite! rewrite plugin to fix numerous issues and add features

### PR DESCRIPTION
**Pull request by tcrawford-figure**
This rewrite fixes many problems that are very apparent in v1.

Breaking changes:
- Plugin `id` change from `com.figure.gradle.semver-plugin` to just `com.figure.gradle.semver`
- The version and version tag output have been moved from `build/semver/version.txt` to `build/semver/semver.properties`

Problems fixed include:
- Inability to build without initializing git repository
- Inability to build when merging, rebasing, reverting, bisecting, cherry-picking, and being in detached head state
- Incorrect version calculation in some scenarios
- Inability to set alternate initial version
- Inability to build without an initial tag

Features added:
- Support for stages
- Support for specifying alternate development and main branch names
- Smart branching strategy inference
- Support for building with build metadata
- Zero config support (in most cases, there's no need to do anything but apply the plugin)
- **Complete documentation to explain how to use the included and optional gradle properties**. Little asterisk: we need to setup GitHub pages for this repository so that documentation can be viewed at https://figuretechnologies.github.io/gradle-semver-plugin

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,242 +1,82 @@
...
+
+At a glance, this plugin provides the following features:
+
+- Support for stages (`rc`, `beta`, `stable`, `snapshot`, etc. - see below)
+- Support for modifiers (`auto`, `patch`, `minor`, `major`)
+- Support for branch-based version calculations
+- Support for overriding the version
+- Support for setting an alternate initial version
+- Support for specifying alternate main and development branch names
+- Support for appending build metadata (format: `+<yyyyMMddHHmmss>`)
```
Comment (jonasg13) [inline README.md:15]: `Support for ...` here seems repetitious to me... Please consider moving it to the above line. Maybe something like this: `At a glance, this plugin supports the following features:` and then just list the items. 🤷🏻‍♂️

File: build.gradle.kts [hunk 2]
```diff
@@ -1,137 +1,271 @@
...
-    prerelease = false
-    repo = "gradle-semver-plugin" // Should automatically pick up your project name, overwrite this if you need something else
-    targetCommitish = System.getenv("TARGET_COMMIT_ISH")?.takeIf { it.isNotBlank() } ?: "main"
+class PublishingConstants {
+    val name = "Gradle Semver Plugin"
+    val description = "Gradle Plugin for Automatic Semantic Versioning"
+    val pluginImplementation = "$group.SemverPlugin"
+    val tags = listOf("semver", "versioning", "git")
+
+    val website = "https://figuretechnologies.github.io/gradle-semver-plugin"
```
Comment (jonasg13) [inline build.gradle.kts:191]: I currently get a 404 for this URL

File: docs/append-build-metadata.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
+# Append Build Metadata
```
Comment (jonasg13) [inline docs/append-build-metadata.md:1]: I like this a lot!

File: docs/index.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+At a glance, this plugin provides the following features:
+
+- Support for stages (`rc`, `beta`, `stable`, `snapshot`, etc. - see below)
+- Support for modifiers (`auto`, `patch`, `minor`, `major`)
+- Support for branch-based version calculations
+- Support for overriding the version
+- Support for setting an alternate initial version
+- Support for specifying alternate main and development branch names
+- Support for appending build metadata (format: `+<yyyyMMddHHmmss>`)
```
Comment (jonasg13) [inline docs/index.md:15]: See previous comment about `Support for ...`

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -1,49 +1,40 @@
...
-
-[libraries]
+kotlin = "1.9.24" # Goal is to track whatever the latest stable version that is supported by Gradle
 
-# Libraries
-eclipse-jgit-eclipseJgit = { module = "org.eclipse.jgit:org.eclipse.jgit", version.ref = "jgit" }
-swiftzer-semver = { module = "net.swiftzer.semver:semver", version.ref = "swiftzer-semver" }
+jgit = "7.0.0.202409031743-r"
+kotest = "5.9.1"
+kotlin-semver = "2.0.0"
```
Comment (jonasg13) [inline gradle/libs.versions.toml:7]: 🐥 || 🥚

File: gradle/libs.versions.toml [hunk 6]
```diff
@@ -1,49 +1,40 @@
...
-[bundles]
+jgit = { module = "org.eclipse.jgit:org.eclipse.jgit", version.ref = "jgit" }
+kotlin-semver = { module = "io.github.z4kn4fein:semver", version.ref = "kotlin-semver" }
 
-kotest = ["kotest-junit5-jvm", "kotest-assertions-core-jvm", "kotest-framework-datatest"]
+kotest-datatest = { module = "io.kotest:kotest-framework-datatest", version.ref = "kotest" }
+kotest-runner = { module = "io.kotest:kotest-runner-junit5-jvm", version.ref = "kotest" }
+testkit-support = { module = "com.autonomousapps:gradle-testkit-support", version.ref = "testkit-support" }
 
 [plugins]
```
Comment (jonasg13) [inline gradle/libs.versions.toml:29]: Is the `signing` plugin builtin?

File: mkdocs.yml [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+copyright: 'Copyright &copy; 2024 Figure Technologies'
+
+theme:
+  name: material
+  language: 'en'
+  favicon: 'assets/package-white.svg'
+  logo: 'assets/package-white.svg'
+  font:
+    code: 'Robot Mono'
```
Comment (jonasg13) [inline mkdocs.yml:18]: I believe the font is called `Roboto Mono` 🤷🏻‍♂️

File: build.gradle.kts (hunk 2 above)
Comment (tcrawford-figure) [reply build.gradle.kts:191]: @jonasg13 you will until this gets merged in and the documentation gets generated on the `gh-pages` branch.

File: gradle/libs.versions.toml (hunk 6 above)
Comment (tcrawford-figure) [reply gradle/libs.versions.toml:29]: Yes. `signing` is a builtin plugin

File: mkdocs.yml (hunk 7 above)
Comment (tcrawford-figure) [reply mkdocs.yml:18]: Omg... 😅 Great catch! I'll have to fix that!

**Comment by tcrawford-figure** [comment]
Please review https://github.com/FigureTechnologies/gradle-semver-plugin/pull/147 since that has ticket associated with the changes.

**Pull request closed without merging**